src/main/java/org/embulk/gradle/embulk_plugins/ProspectiveDependencies.java

```diff
@@ -0,0 +1,368 @@
+                final LinkedHashSet<VersionlessDependency> duplicates,
+                final ArrayList<LibraryBinaryIdentifier> libs,
+                final ArrayList<ComponentIdentifier> others) {
+            super(buildMessage(duplicates, libs, others));
+            this.duplicates = duplicates;
+            this.libs = libs;
+            this.others = others;
+        }
+
+        private static String buildMessage(
+                final LinkedHashSet<VersionlessDependency> duplicates,
+                final ArrayList<LibraryBinaryIdentifier> libs,
+                final ArrayList<ComponentIdentifier> others) {
+            final StringBuilder exceptionMessage = new StringBuilder();
+
+            if (!duplicates.isEmpty()) {
+                exceptionMessage.append("Dependency duplicates are found: ");
+                exceptionMessage.append(
+                        duplicates.stream().map(VersionlessDependency::toString).collect(Collectors.joining(", ", "[", "]")));
+            }
+
+            if (!libs.isEmpty()) {
+                if (exceptionMessage.length() > 0) {
+                    exceptionMessage.append(" ");
+                }
+                exceptionMessage.append("Native library binary dependencies are not supported: ");
+                exceptionMessage.append(
+                        libs.stream().map(LibraryBinaryIdentifier::getDisplayName).collect(Collectors.joining(", ", "[", "]")));
+            }
+
+            if (!others.isEmpty()) {
+                if (exceptionMessage.length() > 0) {
+                    exceptionMessage.append(" ");
+                }
+                exceptionMessage.append("Unknown type of artifacts: ");
+                exceptionMessage.append(
+                        others.stream().map(ComponentIdentifier::getDisplayName).collect(Collectors.joining(", ", "[", "]")));
+            }
+
+            return exceptionMessage.toString();
+        }
+
+        private final LinkedHashSet<VersionlessDependency> duplicates;
+        private final ArrayList<LibraryBinaryIdentifier> libs;
+        private final ArrayList<ComponentIdentifier> others;
+    }
+
+    private static LinkedHashMap<VersionlessDependency, String> buildVersionMapFromResolvedArtifacts(
+            final Set<ResolvedArtifact> artifacts)
+            throws UnexpectedDependencyException {
+        final LinkedHashMap<VersionlessDependency, String> versionMap = new LinkedHashMap<>();
+
+        final LinkedHashSet<VersionlessDependency> duplicates = new LinkedHashSet<>();
+        final ArrayList<LibraryBinaryIdentifier> libs = new ArrayList<>();
+        final ArrayList<ComponentIdentifier> others = new ArrayList<>();
+
+        for (final ResolvedArtifact artifact : artifacts) {
+            final ComponentIdentifier componentIdentifier = artifact.getId().getComponentIdentifier();
+
+            if (componentIdentifier instanceof ProjectComponentIdentifier) {
+                // @@@@
```

Comment:
nitpick, what does `@@@@` mean?

Reply:
Oops, I forgot it, too. Good catch! It'll be `UnsupportedOperationException` for the time being. Will fix.

Reply:
Addressed in 79bbe27606bd4ddcaf89f64d07daea0a00129ef3. Finally mixed into `UnexpectedDependencyException`.